### PR TITLE
Fix /pass bypass security issue.

### PR DESCRIPTION
**Pull request by rdebath**
Mitigation:
    If verify-names is enabled the authentication is handled by the TPP
    that server overrides the username processing.  The default server,
    classicube.net, uses case sensitive names with the initial logon
    mapping the entered userid to the correct case. It is expected that
    alternate servers will also not allow user names that collide under
    case conversion.

Issue:
    The username is usually treated as case insensitive, however,
    the password is stored in a file with the user name in it's name.
    For filesystems that ignore case this will only occasionally an issue
    if unicode case conversion is inconsistent. But linux filesystems
    do not ignore case so all letters (and letter like characters)
    in the username have two representations.

Contributory factor:
    If the password file for a particular userid is not found the user
    is prompted to enter a new password without any authenticatin.

Other notes:
It may also be a good idea to use a conversion to lowercase when creating the salt for the password file from the username. As it currently stands if a user enters the wrong case for their username they will get a wrong password prompt rather than the previous issue of having to use /setpass without authentication.

### CODE REVIEW

**Comment by UnknownShadow200** [comment]
Hmm, I don't think always iterating through all the files in the directory is the best solution. In my opinion, this should instead be done as:
a) Always convert username to lowercase when writing password file
b) When reading, try to read password file with username converted to lowercase first. If the file doesn't exist, then try to read the password file without converting username to lowercase

---
As far as I know, there's very few servers out there without verify names and purely using /pass. So I think it's reasonable to require them to manually rename password files, or have the server do that once automatically on startup

**Comment by rdebath** [comment]
Remember the username is used as the salt for the password file so they cannot be renamed.

This means that your method would break logins any windows install as the 'name converted to lowercase' would succeed to open the file, but the hash would be tied to the case insensitive version of the user name and give a "wrong password".

In addition your method would not block the security issue, if someone used the wrong case on a user that has not been converted they would get the `/setpass` prompt and take over the lowercased file.

I suppose you could change the extension (or directory) for the new files and so use the directory scan only for old files then create a new file (with new extension) and delete the old one when the user next logs in (and you have the unhashed password available).

Of course if you're going to replace the files, it might also be a good idea link the password to the player record's id. 

****! The player table is case sensitive too, looks like you might need some "COLLATE NOCASE" tweaks.

At least this might be useful:
`select * from players where name not in (select name from players group by name COLLATE NOCASE) ;`

**Comment by UnknownShadow200** [comment]
Good point. Maybe instead automatically convert all password files to lowercase at startup. When reading, compute two hashes for lowercase username and current username, then check if either hash equals

----

The player's name is case sensitive in a number of other places too. Player /nick, /inbox, /ignore list, etc

**Comment by rdebath** [comment]
Yes, but there's no exit plan there, you won't know when everybody (important) has been updated.

Okay, I've updated the patch to do a full re-hash of the passwords using the normal methods.

In addition because of this I've taken the opportunity to fix two other small issues; specifically, (a) you shouldn't really be using MD5 anymore, while it's not yet a security issue it needs to be changed. And (b) non-ascii characters in the password are changed to "?" <s>so a password like "пароль" is identical to "??????"</s>.

This update happens when someone uses the `/pass` command.
If a new format file is found it is used directly; I've changed the extension to `.pwd` so these can be instantly identified.
If no new format file is found I search for matching `.dat` files if one is found (and the hash matches) the old file is removed and a new file is created using the just entered password.

After a little while all the files for users who logon regularly will be in the new format and so be fast.
At this point you can disable any users who don't have new files. (eg: create broken new files for them manually)

Note you can also turn on the `DISABLE_OLD_PASSWORDFILE` compile flag to turn off the old method completely. That also flags the old code to be deleted at some point.

Just one thing; is there anything else that it would be good to change in the password files?

**Comment by Goodlyay** [comment]
Might be worth noting that a player would never be able to set their password to "пароль" because ClassiCube only supports code page 437

**Comment by rdebath** [comment]
Under linux it supports unicode.

**Comment by Goodlyay** [comment]
I'd be interested in seeing the client send unicode to the server, don't think that's possible.

**Comment by rdebath** [comment]
Humm, I can't paste into the client and the only non-ascii key I have is "£", which is in CP437 and gets flattened into "?".
Nope, wrong: I can paste, but it only accepts CP437 characters (one by one) presumably because it converts them.

So you're right (obviously :grin:) but I don't think that changes much as flattening even the CP437 characters is a bad idea.

**Comment by UnknownShadow200** [comment]
Still not sure why this can't be one time conversion of filenames to lower case. And then when reading, first try to check new hash computation with lowercase username. And if that fails, fallback on old hash computation with just username.

**Comment by rdebath** [comment]
Okay
* There's no exit plan, it's difficult to tell who's password has not been converted so you can't tell which accounts to disable.
* You'd be deleting the salt, ie the userid that created the file, this means you cannot rollback the change.
* You're deleting or making inaccessible password files for anyone that's been attacked so you've got a better than 50% chance of locking them out. This is especially true of high value accounts like SuperOPs.

Remember also that SuperOPs are likely to be logging in from a Private IP and so bypass the name validation. There's a good chance that they've left the password turned on for their login even if it's off for everyone else. (Paranoia works too)

Pissed off SuperOP ⟶ 👿🔊

I've also added a small update (It needs testing) which should avoid the scan for all but first login ever without locking anyone out who belongs there.

**Comment by UnknownShadow200** [comment]
Has there even been a server attacked like this before?

**Comment by rdebath** [comment]
I expect so, but we can't tell. 

If the superop gets another set-password prompt while they're learning they just assume flaky software. Once they've entered their password a couple of times they'll stop getting the message for all variants of the user name they use. 

If a site is destroyed from outside the OPs would have a scapegoat in terms of the user who's been cloned so again you wouldn't hear about it. If the Op cloned is trusted they might suspect something, but they'd have to prove an attack route (like I have done) to expect to be listened to.

In summary, it's easy to find by accident so it's probably happened, but usually the results are trivial because it's the user themselves who "attack" their own account. Even if it's a malicious attack the cloned user would get blamed so there won't be a github issue.

PS: Also we're being watched  👀

**Comment by UnknownShadow200** [comment]
My rationale is that 
1.  The number of users who play without name verification, on a server hosted on linux, and have differing passwords for multiple case variations of their username is likely to be very small
2. I'd rather the server loudly shows an error on startup about failing to move a password file, to make it obvious that there is an issue with passwords that needs to be addressed

FYI, wouldn't bother with adding a macro to compile time remove 'old password' support, because supporting old passwords will have to be kept around anyways for backwards compatibility. 
Also don't need to preserve the `>` to `)` conversion in `ComputeHash` - that was only kept around for backwards compatibility.
Don't see much point of adding `MCGalaxy: ` prefix in `ComputeHash` though, because the unique username is already being prefixed so you can't just lookup the hash.

**Comment by rdebath** [comment]
1. I am assuming the passwords are the same so that people don't look into the so called flakyness. If you pick the wrong casing of the user name when you rename the file the user will be locked out because the MD5 generation is case sensitive on both the user and password.
2. This is likely to be an automatic or semi-automatic update so the restart messages won't be seen no matter how obnoxious you make them. In the event that there is a problem the owner's most reasonable course of action is to rollback to the working version ... which won't work. As for an active attack; you said it yourself, it's not going to be common and the normal response of locking/demoting a pwn'd account does exactly the right thing.

The '(', '<' conversion, yes, I wondered about that I'll remove it.

Old passwords; I would suggest about a year, probably two for backwards compatibility. After that it's reasonable to consider a password as expired and need a reset by console or other user.

The prefix; No, most definitively. A completely unique per password file salt is what should be used, this forces password cracking to be done just for this instance. Using a host specific (or application specific) salt isn't as good but does mean that a precomputation attack has to be done for this specific server or application. Just the username very poor salt, that are only a few thousand possibilities and some of them are very, very common across lots of applications. Rainbow tables do exist for common username/password pairs, even Wikipedia mentions them, but just adding a constant string makes them a lot less useful. Though putting the string on the end might be better if you want to slow down someone making a table.

Note: There probably isn't a rainbow table for your current (weird) method of triple hashing, but hashing the concatenation is the standard method so you need a stronger salt, or some different weirdness like the prefix. How about making the string "0bec662b-416f-450c-8f50-664fd4a41d49" ? Then it's more obvious that it's a security hack.

**Comment by UnknownShadow200** [comment]
duuuhhh I'm dumb, of course it won't work with differently cased names.

Sure, fine to change the string to that

**Comment by UnknownShadow200** [comment]
Thanks. Apologies for wasted time due to my misunderstanding
